Implement read-write (shared) lock from ZooKeeper recipe.

cpatulea opened the pull request:
Reference: http://zookeeper.apache.org/doc/trunk/recipes.html#Shared+Locks

This is a writer-preference shared lock.

Most of the lock protocol is the same as for an exclusive lock, except that it needs to be parameterized on:
1) Node name - lock "kind" - a write lock (**lock**) or read lock (**rlock**).
2) Which node names block which other node names (read locks block only writers, writers block both).
3) Finding the predecessor node according to the rule in #2.

These changes are integrated to the main Lock class. The changes are backward compatible with existing users of Lock. Users should normally never pass the Lock node_name and exclude_names kwargs directly, they should use WLock and RLock.

cpatulea commented:
/cc @pphaneuf @kamalmarhubi FYI; if you have time/interest to review that would also be appreciated :)

kamalmarhubi commented on kazoo/recipe/lock.py:
```diff
@@ -54,7 +54,13 @@ def __init__(self, client, path, identifier=None):
+        :param exclude_names: Node names which exclude this contender when
+                              present at a lower sequence number. Involved in
+                              read/write locks. For a normal (exclusive) lock,
+                              leave unset.
```
I'd consider moving the comment that these are for read/write locks to the front, so that uninterested people know they can ignore them.

cpatulea replied:
Done.

kamalmarhubi commented on kazoo/recipe/lock.py:
```diff
@@ -65,11 +71,19 @@ def __init__(self, client, path, identifier=None):
 
+        if node_name is None:
+            node_name = "__lock__"
+        self.node_name = node_name
```
strings are immutable, so you could safely put the default in the arglist

cpatulea replied:
Done.

kamalmarhubi commented on kazoo/recipe/lock.py:
```diff
@@ -44,9 +44,9 @@ class Lock(object):
     acquired will raise a ``RuntimeError``.
 
     """
-    _NODE_NAME = '__lock__'
```
maybe leave this in as _DEFAULT_NODE_NAME?

cpatulea replied:
Eh. This also gets used in WLock and RLock, so module scope would be a more appropriate place.

But I'm not sure there's a point in defining symbolic constants for these anyway. They're already just arbitrary tokens, they're readily searchable, and they're unlikely to ever change, so I figure 'explicit' in this case would aid readability.

kamalmarhubi commented on kazoo/recipe/lock.py:
```diff
@@ -44,9 +44,9 @@ class Lock(object):
     acquired will raise a ``RuntimeError``.
 
     """
```
Do you want to add mention of the reader/writer lock in the class docstring?

cpatulea replied:
Done.

kamalmarhubi commented on kazoo/recipe/lock.py:
```diff
@@ -267,6 +291,56 @@ def __exit__(self, exc_type, exc_value, traceback):
+    .. code-block:: python
+
+        zk = KazooClient()
+        lock = WLock("/lockpath", "my-identifier")
```
Obviously I don't know the code base, but the creation and non-use of `zk` looks fishy to me. Is there an implicit global client?

cpatulea replied:
Thanks, done.

harlowja replied:
That shouldn't be the case (no global), probably just the docs are not right here.

kamalmarhubi commented:
High level comment: it looks as though nodes following the read/write lock protocol could interact with nodes following the exclusive lock protocol at the same node name. This could lead to safety issues since the exclusive lock protocol doesn't exclude `__rlock__`: a node following the exclusive protocol would happily take the lock even if there is a reader, and go on its merry way.

I may be wrong, in a morning code review rush. :-)

harlowja commented on kazoo/recipe/lock.py:
```diff
@@ -267,6 +291,56 @@ def __exit__(self, exc_type, exc_value, traceback):
+
+    .. code-block:: python
+
+        zk = KazooClient()
```
Same issue as above, doesn't exactly look right.

harlowja commented on kazoo/recipe/lock.py:
```diff
@@ -167,14 +181,15 @@ def _inner_acquire(self, blocking, timeout):
 
             # otherwise we are in the mix. watch predecessor and bide our time
-            predecessor = self.path + "/" + children[our_index - 1]
+            predecessor = self.path + "/" + predecessor
```
While your at it how about u use the following:

https://github.com/python-zk/kazoo/blob/master/kazoo/protocol/paths.py#L19

harlowja replied:
Good point, let's wait for those.

harlowja commented:
Have u seen https://github.com/python-zk/kazoo/pull/141 (might be an intersting thing to look at for u).

harlowja commented on kazoo/recipe/lock.py:
```diff
@@ -43,18 +43,31 @@ class Lock(object):
-    _NODE_NAME = '__lock__'
 
-    def __init__(self, client, path, identifier=None):
+    def __init__(self, client, path, identifier=None, node_name="__lock__",
```
I'd almost rather have ``class Lock(object):`` -> ``class _Lock(object):``

And then only have these keyword arguments on ``_Lock``; what do u think? Even though u are saying 'don't pass them' it seems like we could do better...

cpatulea commented:
#141 looks nice, I would be happy to use it.

Are you willing to merge it?
On Apr 7, 2015 5:20 PM, "Joshua Harlow" <notifications@github.com> wrote:

> Have u seen #141 <https://github.com/python-zk/kazoo/pull/141> (might be
> an intersting thing to look at for u).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/python-zk/kazoo/pull/306#issuecomment-90735916>.
>

harlowja commented:
#141 could be combined with yours; anything useful from it that u might want?

cpatulea commented:
My application code can work as well on top of #141 or this pull request.

I'm also happy to make some minor changes as needed to either #141 or this pull request, so that they can be merged.

However, I feel that this pull request is getting sidetracked. I sent the pull request nearly a week ago, I've addressed all review comments so far, and you are now suggesting some kind of merging of the pull requests, which would involve significant additional work. I am happy to help improve kazoo, but at this point either pull request seems like a good candidate for merging, without additional revision. Any further work can be done later, by whoever depends on that work.

harlowja commented:
Fair enough, I didn't mean to discredit your code or anything; was just offering some ideas...

OpenSource takes time (and patience); but the more you participate the better it gets :)

harlowja commented on kazoo/recipe/lock.py:
```diff
@@ -195,8 +218,14 @@ def _get_sorted_children(self):
         # can't just sort directly: the node names are prefixed by uuids
-        lockname = self._NODE_NAME
-        children.sort(key=lambda c: c[c.find(lockname) + len(lockname):])
+        def _seq(c):
```
It'd be nice to have a comment here ;)

cpatulea replied:
Done.

harlowja commented:
Ok let's get this in, further tweaks can be made after merge I think :+1:

bbangert commented on kazoo/recipe/lock.py:
```diff
@@ -267,6 +303,56 @@ def __exit__(self, exc_type, exc_value, traceback):
         self.release()
 
 
+def WLock(*args, **kwargs):
```
Why aren't these subclasses as is typical in Python to slightly modify behavior? The lack of subclassing here means that when you look at tracebacks and such, there's nothing indicating WLock vs RLock vs Lock. That will drive people crazy.

bbangert commented:
This recipe unfortunately creates Lock objects that are actually Read or Write Locks, without anything in the class object to easily identify them as such. Given the scope of the changes, it definitely needs a thorough set of tests as well which don't seem to be included.

jeblair commented:
I have continued work on this PR in #419.

bbangert commented:
Ok, closing this one then.
